pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,399 @@
+	pendingCRUpdatesObj                      *pendingCRUpdates
+	k8sClient                                clientset.Interface
+	controllerRuntimeClient                  client.Client
+)
+
+// StartSvFSSReplicationService Starts SvFSSReplicationService
+func StartSvFSSReplicationService(ctx context.Context, svFeatureStatConfigMapName string, svFeatureStateConfigMapNamespace string) error {
+	log := logger.GetLogger(ctx)
+	log.Info("starting SvFSSReplication service")
+
+	supervisorFeatureStatConfigMapName = svFeatureStatConfigMapName
+	supervisorFeatureStateConfigMapNamespace = svFeatureStateConfigMapNamespace
+	pendingCRUpdatesObj = &pendingCRUpdates{
+		lock:            &sync.RWMutex{},
+		pendingCRUpdate: make(map[string]*featurestatesv1alpha1.CnsCsiSvFeatureStates),
+	}
+	// This is idempotent if CRD is pre-created then we continue with initialization of svFSSReplicationService
+	err := k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		reflect.TypeOf(featurestatesv1alpha1.CnsCsiSvFeatureStates{}).Name(), CRDGroupName, internalapis.SchemeGroupVersion.Version, apiextensionsv1beta1.NamespaceScoped)
+	if err != nil {
+		log.Errorf("failed to create CnsCsiSvFeatureStates CRD. Error: %v", err)
+		return err
+	}
+
+	// Create the kubernetes client
+	k8sClient, err = k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("create k8s client failed. Err: %v", err)
+		return err
+	}
+	// get kube config
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Error: %v", err)
+		return err
+	}
+	// create controller runtime client
+	controllerRuntimeClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("failed to create controllerRuntimeClient. Err: %v", err)
+		return err
+	}
+
+	// create/update feature state CRs in all workload namespaces
+	err = updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, []featurestatesv1alpha1.FeatureState{})
+	if err != nil {
+		log.Errorf("failed to update feature states CR in the all workload namespaces. err: %v", err)
+		return err
+	}
+
+	// Create k8s Informer and watch on configmaps and namespaces
+	informer := k8s.NewInformer(k8sClient)
+	informer.AddConfigMapListener(ctx, k8sClient, svFeatureStateConfigMapNamespace,
+		// Add
+		nil,
+		// Update
+		func(oldObj interface{}, newObj interface{}) {
+			configMapUpdated(oldObj, newObj)
+		},
+		// Delete
+		nil)
+	informer.AddNamespaceListener(
+		// Add
+		func(obj interface{}) {
+			namespaceAdded(obj)
+		},
+		// Update
+		nil,
+		// Delete
+		nil)
+	informer.Listen()
+	log.Infof("informer on config-map and namespaces started")
+
+	// Start Retry routine to process failed create or update CRs
+	go pendingCRUpdatesObj.startRetryAttempts()
+	log.Infof("started background routine to process failed CR updates at regular interval")
+	log.Infof("SvFSSReplicationService is running")
+	var stopCh = make(chan bool)
+	<-stopCh
```

Comment:
What happens if we do not block?

Reply:
if we do not block, then routine will end.

Reply:
I meant to ask, would that be a problem? And moreover we have a local channel here which means another go routine cannot stop this go routine if needed (ex: graceful shutdown of syncer).